src/librustc/middle/dead.rs

```diff
@@ -445,6 +445,11 @@ impl<'a, 'tcx> DeadVisitor<'a, 'tcx> {
             && !has_allow_dead_code_or_lang_attr(&variant.attrs)
     }
 
+    fn should_warn_about_foreign_item(&mut self, fi: &hir::ForeignItem) -> bool {
+        !self.symbol_is_live(fi.id, None)
+            && !has_allow_dead_code_or_lang_attr(&fi.attrs)
```

Comment:
Pre-existing but why do we snoop lint attributes instead of registering the lints and letting the lint pass track them? cc @rust-lang/compiler